### PR TITLE
improvement: Improving ccm performance

### PR DESCRIPTION
**Pull request by Orenef11**
* Remove sleep and use `wait_for` method
* Reduce most sleep to 50 ms
* Bug ticket https://github.com/scylladb/scylla-dtest/issues/2429

### CODE REVIEW

File: ccmlib/node.py
Comment (Orenef11) [inline ccmlib/node.py:733]: This function can run indefinitely ...
I limited the duration to 24 hours.

**Comment by roydahan** [comment]
Do not merge this until we do the move to pytest.
I don't want new instabilities, don't want to do more runs.
We will do the migration, make sure all runs exactly like in the past and then merge optimizations.

**Review by nyh — COMMENTED** [review]
I left some comments, but the biggest problem I had reviewing this is that I didn't know how you decided which of the timeouts to change. When is each of these different timeout relevant? Did you check that each of these changes improved something in some operation (starting a cluster, shuttting it down, or what?). A lot of lowering of the sleeps may really improve something, but I just don't know what.

File: ccmlib/scylla_node.py
Comment (Orenef11) [reply ccmlib/scylla_node.py]: Yes, it's too short but I can increase this number (now the value is 500 ms).
Again, the CPU time is wasted anyway.

Yes, this method call from `ScyllaNode.start --> _start_scylla -> wait_for_starting (if wait_for_binary_proto is  True)

File: ccmlib/node.py
Comment (Orenef11) [inline ccmlib/node.py:734]: @nyh 
I can't remove the `timeout` variable because it's required parameter in `wait_for` method

File: ccmlib/scylla_node.py
Comment (nyh) [reply ccmlib/scylla_node.py]: It turns out this change is very important - see https://github.com/scylladb/scylla-dtest/issues/2429#issuecomment-994817655
Without this change, dtest simply sleeps 10 seconds - for no reason whatsoever - after `cluster.start()` completes!

**Comment by nyh** [comment]
> Do not merge this until we do the move to pytest. I don't want new instabilities, don't want to do more runs. We will do the migration, make sure all runs exactly like in the past and then merge optimizations.

If @roydahan doesn't want the full patch, @Orenef11 please consider splitting **small patches** out of this big one, each fixing just one thing which we do understand and did test. The one improvement in this patch which I would recommend to split off first - and I was able to reproduce and measure  was the change of `sleep_time = 10 if timeout >= 100 else 1` to `sleep_time = 0.5`. In my test starting a 5-node cluster, this 10-second sleep caused a useless 10-second sleep _after_ `cluster.start()` finished. If your patch only fixed this (and doesn't do any other refactoring), I don't see how @roydahan could be against it. Then later you can have a PR which does the rest of the changes that you propose.

Even if parts or even most of this patch are held off to later, let's not hold of the parts we already tested and can prove - namely that sad 10-second sleep after the cluster is created. @roydahan, for too many years (!) we let dtest be much slower than it needs to be - it was so slow that nobody even noticed a useless 10 second sleep. Every time someone suggested to improve this, we kept postponing it - "pytest will solve it", "raft will solve it". In fact, we can already solve most of the slowness today, without pytest or raft or any big refactorings, just tiny patches. So let's do it. Today starting a 5-node cluster in dtest takes 224 seconds. With the fixes we are proposing (not requiring this full patch, just the "sleep_time = 10" part), this number goes down to 9 seconds. That's **25 times faster**. Let's embrace this opportunity.

**Comment by fruch** [comment]
> > Do not merge this until we do the move to pytest. I don't want new instabilities, don't want to do more runs. We will do the migration, make sure all runs exactly like in the past and then merge optimizations.
> 
> If @roydahan doesn't want the full patch, @Orenef11 please consider splitting **small patches** out of this big one, each fixing just one thing which we do understand and did test. The one improvement in this patch which I would recommend to split off first - and I was able to reproduce and measure was the change of `sleep_time = 10 if timeout >= 100 else 1` to `sleep_time = 0.5`. In my test starting a 5-node cluster, this 10-second sleep caused a useless 10-second sleep _after_ `cluster.start()` finished. If your patch only fixed this (and doesn't do any other refactoring), I don't see how @roydahan could be against it. Then later you can have a PR which does the rest of the changes that you propose.
> 
> Even if parts or even most of this patch are held off to later, let's not hold of the parts we already tested and can prove - namely that sad 10-second sleep after the cluster is created. @roydahan, for too many years (!) we let dtest be much slower than it needs to be - it was so slow that nobody even noticed a useless 10 second sleep. Every time someone suggested to improve this, we kept postponing it - "pytest will solve it", "raft will solve it". In fact, we can already solve most of the slowness today, without pytest or raft or any big refactorings, just tiny patches. So let's do it. Today starting a 5-node cluster in dtest takes 224 seconds. With the fixes we are proposing (not requiring this full patch, just the "sleep_time = 10" part), this number goes down to 9 seconds. That's **25 times faster**. Let's embrace this opportunity.

FYI no one said pytest would solve any of this, just that we are at a point changing too many things, like this PR is risky without proper testing of running all the tests, and we are in the midst of moving to pytest.

**Comment by roydahan** [comment]
@nyh today we did the switch and now dtest moved to pytest.
Now, whatever optimization we want to introduce should be tested making sure all tests are passing and aren’t flakey.
Also @bhalevy should approve, as he also the one who suffer from instabilities and working hard to stable them.

File: ccmlib/scylla_node.py
Comment (nyh) [reply ccmlib/scylla_node.py]: I don't know how this is possible, but it seems that after the dtest switch to pytest, this 10-second sleep no longer happens - and in my test this code snippet doesn't even run.

**Comment by nyh** [comment]
I tested this patch together with my dtest patch https://github.com/scylladb/scylla-dtest/pull/2455. I didn't see this patch improve anything with boot time (which after 2455 was around 8.8 seconds for 5 nodes, and wasn't reduced further by this patch), and probably didn't help with shutdown time either (but this is harder to call as the shutdown time seems to fluctuate quite a bit).

Perhaps this patch helps other operations, but I don't know how to test it.

File: ccmlib/scylla_node.py
Comment (bhalevy) [inline ccmlib/scylla_node.py:1250]: How did result end up as `None` previously?

This looks like a bug fix on its own, hence be separated and be considered for backport.

**Review by bhalevy — CHANGES_REQUESTED** [review]

**Comment by bhalevy** [comment]
Can this PR be closed?

**Comment by fruch** [comment]
> Can this PR be closed?

yes

**Pull request closed without merging**
